packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -582,22 +606,92 @@ export abstract class SharedObjectCore<
 		this.emitInternal("op", message, local, this);
 	}
 
+	/* eslint-disable jsdoc/check-indentation */
 	/**
-	 * Process messages for this shared object. The messages here are contiguous messages for this object in a batch.
+	 * Process a bunch of messages for this shared object. A bunch is group of messages that have the following properties:
+	 * - They are all part of the same grouped batch, which entails:
+	 *   - They are contiguous in sequencing order.
+	 *   - They are all from the same client.
+	 *   - They are all based on the same reference sequence number.
+	 *   - They are not interleaved with messages from other clients.
+	 * - They are not interleaved with messages from other DDS in the container.
 	 * @param messageCollection - The collection of messages to process.
+	 *
 	 */
+	/* eslint-enable jsdoc/check-indentation */
 	private processMessages(messagesCollection: IRuntimeMessageCollection): void {
-		const { envelope, messagesContent, local } = messagesCollection;
-		for (const { contents, localOpMetadata, clientSequenceNumber } of messagesContent) {
-			this.process(
-				{
-					...envelope,
-					clientSequenceNumber,
-					contents: parseHandles(contents, this.serializer),
-				},
-				local,
+		this.verifyNotClosed(); // This will result in container closure.
+		const { envelope, local, messagesContent: encodedMessagesContent } = messagesCollection;
+
+		// Decode any handles in the contents and emit the "pre-op" event.
+		const decodedMessagesContent: IRuntimeMessagesContent[] = [];
+		for (const { contents, localOpMetadata, clientSequenceNumber } of encodedMessagesContent) {
+			const decodedMessageContent: IRuntimeMessagesContent = {
+				contents: parseHandles(contents, this.serializer),
 				localOpMetadata,
-			);
+				clientSequenceNumber,
+			};
+			decodedMessagesContent.push(decodedMessageContent);
+
+			const decodedMessage: ISequencedDocumentMessage = {
+				...envelope,
+				contents: decodedMessageContent.contents,
+				clientSequenceNumber,
+			};
+			this.emitInternal("pre-op", decodedMessage, local, this);
```

Comment:
@anthony-murphy @vladsud Do you guys know why we have a "pre-op" event and if it's used?

Reply:
Looks like it was added at the beginning of time - https://github.com/microsoft/FluidFramework/commit/f2bcc60286274b6adec6b49bf1fd5a106f19a256. It's not used from what I have checked. I am inclined to remove this event.